### DISCUSSION TITLE
Pinned-Dependencies

### DISCUSSION THREAD

**Discussion by stephenegriffin** [ossf/scorecard#4422] [Q&A] [answered]
Getting a code scanning alert in https://github.com/microsoft/MHA: "npmCommand not pinned by hash"
![image](https://github.com/user-attachments/assets/0edee06f-f225-48af-a340-bc842a532c07)

The remediations steps don't spell out what "npmCommand not pinned by hash" or how I'm supposed to resolve it. I basically have no idea what I'm supposed to do here. I tried altering the npm command (https://github.com/microsoft/MHA/pull/1338) but that had no effect.

This issue looks promising: https://github.com/ossf/scorecard/issues/2911 but I don't see any evidence of parsing failures - it would be helpful if I could see any logs showing how this alert was generated but I don't know where/how to find those.

**Comment by spencerschrock** [comment, marked as answer]
> The remediations steps don't spell out what "npmCommand not pinned by hash" or how I'm supposed to resolve it. I basically have no idea what I'm supposed to do here

Agreed, the resolution steps aren't clear in the output.

In this case, it's looking for [`npm ci`](https://docs.npmjs.com/cli/v8/commands/npm-ci) and not `npm i`: 
```dockerfile
RUN cd /app/MHA && npm ci && npm run build --if-present
```

Please double check your Dockerfile still works like this, but the warning in Scorecard should go away (`--local` lets you point Scorecard at local files so you don't need to merge and test):

```shell
scorecard --local=/tmp/MHA --checks Pinned-Dependencies --format json --show-details | jq

# omitted ...

      "details": [
        "Info:  21 out of  21 GitHub-owned GitHubAction dependencies pinned",
        "Info:  12 out of  12 third-party GitHubAction dependencies pinned",
        "Info:   1 out of   1 containerImage dependencies pinned",
        "Info:   1 out of   1 npmCommand dependencies pinned"
      ],
      "score": 10,
      "reason": "all dependencies are pinned",
      "name": "Pinned-Dependencies",
```
